XamlC error : Value cannot be null. (Parameter 'key')

### Description

Using this code:

https://github.com/mattleibow/scorecard-day-nft-marketplace/tree/issues/xamlc-null-key

And this command:

```
 *  Executing task: dotnet build -t:Run -p:Configuration=Debug -f net9.0-android -p:AdbTarget=-s%20emulator-5554 -p:AndroidAttachDebugger=true -p:AndroidSdbTargetPort=62984 -p:AndroidSdbHostPort=62984 -p:CustomAfterMicrosoftCSharpTargets="c:\Users\maleib\.vscode\extensions\ms-dotnettools.dotnet-maui-1.3.3-win32-x64\dist\resources\Custom.After.Microsoft.CSharp.targets" -p:MauiVSCodeBuildOutputFile="C:\Users\maleib\AppData\Local\Temp\dotnet-maui\maui-vsc-02d574b0-4193-4d89-99d1-f44c0a215bea.json" -p:AndroidSdkDirectory="C:\Users\maleib\android-sdk" '-p:JavaSdkDirectory="C:\Program Files (x86)\Android\openjdk\jdk-17.0.8.101-hotspot"' -p:XamlTools="c:\Users\maleib\.vscode\extensions\ms-dotnettools.csharp-2.44.19-win32-x64\.xamlTools" D:\GitHub\scorecard-day-nft-marketplace\NFTMarketplaceApp\NFTMarketplaceApp.csproj 

You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
  NFTMarketplaceApp net9.0-android failed with 1 error(s) and 4 warning(s) (2,7s)
    D:\GitHub\scorecard-day-nft-marketplace\NFTMarketplaceApp\ExplorePage.xaml.cs(13,16): warning CS8618: Non-nullable property 'Name' must contain a non-null value when exiting constructor. Consider adding the 'required' modifier or declaring the property as nullable.
    D:\GitHub\scorecard-day-nft-marketplace\NFTMarketplaceApp\ExplorePage.xaml.cs(14,16): warning CS8618: Non-nullable property 'Type' must contain a non-null value when exiting constructor. Consider adding the 'required' modifier or declaring the property as nullable.
    D:\GitHub\scorecard-day-nft-marketplace\NFTMarketplaceApp\ExplorePage.xaml.cs(16,16): warning CS8618: Non-nullable property 'Currency' must contain a non-null value when exiting constructor. Consider adding the 'required' modifier or declaring the property as nullable.
    D:\GitHub\scorecard-day-nft-marketplace\NFTMarketplaceApp\ExplorePage.xaml.cs(17,16): warning CS8618: Non-nullable property 'CurrencyName' must contain a non-null value when exiting constructor. Consider adding the 'required' modifier or declaring the property as nullable.
    D:\GitHub\scorecard-day-nft-marketplace\NFTMarketplaceApp\ExplorePage.xaml : XamlC error : Value cannot be null. (Parameter 'key')

Build failed with 1 error(s) and 4 warning(s) in 3,1s
```

I get this error:

>     D:\GitHub\scorecard-day-nft-marketplace\NFTMarketplaceApp\ExplorePage.xaml : XamlC error : Value cannot be null. (Parameter 'key')

### Steps to Reproduce

Run the command or just open in VSCode and run.

### Link to public reproduction project repository

https://github.com/mattleibow/scorecard-day-nft-marketplace/tree/issues/xamlc-null-key

### Version with bug

9.0.0-preview.7.24407.4

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

I was *not* able test on other platforms

### Affected platform versions

_No response_

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

--- a/src/Controls/src/Build.Tasks/SetPropertiesVisitor.cs
+++ b/src/Controls/src/Build.Tasks/SetPropertiesVisitor.cs
@@ -1705,6 +1705,9 @@ namespace Microsoft.Maui.Controls.Build.Tasks
 
 		static bool CanAddToResourceDictionary(VariableDefinition parent, TypeReference collectionType, ElementNode node, IXmlLineInfo lineInfo, ILContext context)
 		{
+			if (parent is null)
+				return false;
+
 			if (collectionType.FullName != "Microsoft.Maui.Controls.ResourceDictionary"
 				&& !collectionType.InheritsFromOrImplements(context.Cache, context.Module.ImportReference(context.Cache, ("Microsoft.Maui.Controls", "Microsoft.Maui.Controls", "ResourceDictionary"))))
 				return false;

--- a/src/Controls/tests/Xaml.UnitTests/Issues/Maui24472.xaml
+++ b/src/Controls/tests/Xaml.UnitTests/Issues/Maui24472.xaml
@@ -1,0 +1,47 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
+             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+             xmlns:local="clr-namespace:Microsoft.Maui.Controls.Xaml.UnitTests"
+             x:Class="Microsoft.Maui.Controls.Xaml.UnitTests.Maui24472">
+    <!-- 
+        This test reproduces issue #24472 where XamlC throws 
+        "Value cannot be null. (Parameter 'key')" when compiling XAML 
+        with x:Array containing custom objects and a DataTemplate with a nested view.
+        
+        The issue is triggered by the combination of:
+        1. x:Array with items in CollectionView.ItemsSource
+        2. DataTemplate in ItemTemplate
+        3. CollectionView.Header with nested content
+    -->
+    <CollectionView>
+        <CollectionView.ItemsSource>
+            <x:Array Type="{x:Type local:Maui24472Item}">
+                <local:Maui24472Item Name="Test1" />
+                <local:Maui24472Item Name="Test2" />
+                <local:Maui24472Item />
+                <local:Maui24472Item />
+            </x:Array>
+        </CollectionView.ItemsSource>
+        <CollectionView.Header>
+            <Grid RowDefinitions="Auto,*">
+                <Label Text="Header" />
+                <Border Grid.Row="1" StrokeShape="RoundRectangle 30">
+                    <Grid>
+                        <Image Source="dotnet_bot.png" />
+                        <Button Text="Test" />
+                    </Grid>
+                </Border>
+            </Grid>
+        </CollectionView.Header>
+        <CollectionView.ItemTemplate>
+            <DataTemplate >
+                <Border StrokeThickness="1">
+                    <Grid ColumnDefinitions="Auto,*,*">
+                        <Border Background="#AA333333" WidthRequest="40" HeightRequest="40" Grid.RowSpan="2" />
+                        <Label Text="{Binding Name}" Grid.Column="1" />
+                    </Grid>
+                </Border>
+            </DataTemplate>
+        </CollectionView.ItemTemplate>
+    </CollectionView>
+</ContentPage>

--- a/src/Controls/tests/Xaml.UnitTests/Issues/Maui24472.xaml.cs
+++ b/src/Controls/tests/Xaml.UnitTests/Issues/Maui24472.xaml.cs
@@ -1,0 +1,28 @@
+using NUnit.Framework;
+
+namespace Microsoft.Maui.Controls.Xaml.UnitTests;
+
+// Simple item class for the test
+public class Maui24472Item
+{
+	public string Name { get; set; }
+}
+
+public partial class Maui24472 : ContentPage
+{
+	public Maui24472() => InitializeComponent();
+
+	[TestFixture]
+	class Tests
+	{
+		[Test]
+		// https://github.com/dotnet/maui/issues/24472
+		// XamlC was throwing "Value cannot be null. (Parameter 'key')" when compiling
+		// XAML with x:Array containing custom objects in CollectionView.ItemsSource
+		public void XArrayInCollectionViewItemsSourceDoesNotCrash([Values] XamlInflator inflator)
+		{
+			var page = new Maui24472(inflator);
+			Assert.That(page, Is.Not.Null);
+		}
+	}
+}

